discovery/discovery_test.go

```diff
@@ -83,3 +84,48 @@ func (s *mockSyncer) Sync(tgs []*config.TargetGroup) {
 		s.sync(tgs)
 	}
 }
+
+type mockTargetProvider struct {
+	callCount *int
+}
+
+func (tp mockTargetProvider) Run(ctx context.Context, up chan<- []*config.TargetGroup) {
+	*tp.callCount = *tp.callCount + 1
```

Comment:
You've got a race here

Reply:
I used an atomic update operation here. https://github.com/prometheus/prometheus/pull/3404/commits/fb7a6a758d81776b14fb24168916e1357d603bfc

Reply:
ping @brian-brazil